File: api/aichat.js
// Vercel Serverless Function: api/aichat.js
// Simple rule-based AI chat handler for Vercel (Node.js)

const RATE_LIMIT_WINDOW = 60 * 1000; // 60s
const RATE_LIMIT_MAX = 10; // max requests per window per IP

// In-memory store: { ip: [timestamp,...] }
const requests = {};

const CHAT_RULES = {
  '你好': '你好！我是 CWJ AI 助手，有什么可以帮你的吗？',
  'hello': "Hello! I'm CWJ AI Assistant. How can I help you?",
  '你是谁': '我是 CWJ Tools 内置的 AI 聊天机器人。',
  '谢谢': '不客气！',
  '再见': '再见！祝你一天愉快！'
};

function sanitize(text, maxLength = 500) {
  if (!text || typeof text !== 'string') return null;
  let t = text.trim();
  if (t.length === 0 || t.length > maxLength) return null;
  // basic escaping for HTML to avoid reflected XSS when used in frontend
  return t.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;');
}

function isRateLimited(ip) {
  const now = Date.now();
  if (!requests[ip]) requests[ip] = [];
  // keep only timestamps within window
  requests[ip] = requests[ip].filter(ts => now - ts < RATE_LIMIT_WINDOW);
  if (requests[ip].length >= RATE_LIMIT_MAX) return true;
  requests[ip].push(now);
  return false;
}

function generateReply(text) {
  const lower = text.toLowerCase();
  for (const k in CHAT_RULES) {
    if (lower.includes(k)) return CHAT_RULES[k];
  }
  const samples = [
    `你问的是「${text}」，这是个好问题！我可以帮你把问题整理成搜索关键词。`,
    `关于「${text}」，我建议查阅相关文档或给出更多细节。`,
    `我理解你在问「${text}」，可以试着把问题分成更小的步骤。`
  ];
  return samples[Math.floor(Math.random() * samples.length)];
}

module.exports = async (req, res) => {
  // Allow CORS for same-origin or any origin if needed
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET,POST,OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

  if (req.method === 'OPTIONS') {
    return res.status(200).end();
  }

  const ip = req.headers['x-forwarded-for'] || req.socket.remoteAddress || 'unknown';
  if (isRateLimited(ip)) {
    return res.status(429).json({ error: '请求太频繁，请稍后再试' });
  }

  if (req.method !== 'POST') {
    return res.json({ message: 'CWJ Tools AI chat API. Use POST with JSON {"text":"..."}.' });
  }

  let body = req.body;
  // Vercel parses JSON automatically; ensure content-type handled
  if (!body) {
    try {
      body = await new Promise((resolve, reject) => {
        let data = '';
        req.on('data', chunk => data += chunk);
        req.on('end', () => resolve(JSON.parse(data || '{}')));
        req.on('error', reject);
      });
    } catch (e) {
      return res.status(400).json({ error: '无法解析请求体' });
    }
  }

  const raw = body.text || '';
  const text = sanitize(raw);
  if (!text) {
    return res.status(400).json({ error: '请输入有效问题（1-500字符）' });
  }

  try {
    const reply = generateReply(text);
    return res.json({ reply });
  } catch (e) {
    console.error(e);
    return res.status(500).json({ error: '服务器错误，请稍后重试' });
  }
};

File: AI_Center/aichat.py
import os
import time
import random
import logging
from collections import defaultdict
from markupsafe import escape
from flask import Flask, render_template, request

# Basic logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

BASE_DIR = os.path.dirname(__file__)
app = Flask(__name__, template_folder=os.path.join(BASE_DIR, 'templates'))

# Simple in-memory rate limiter per IP
request_counts = defaultdict(list)

def is_rate_limited(ip, max_requests=10, window=60):
    now = time.time()
    request_counts[ip] = [t for t in request_counts[ip] if now - t < window]
    if len(request_counts[ip]) >= max_requests:
        return True
    request_counts[ip].append(now)
    return False

# Simple sanitize
def sanitize_input(text, max_length=500):
    if not text:
        return None
    text = text.strip()
    if len(text) > max_length:
        return None
    return escape(text)

# Rule-based replies
CHAT_RULES = {
    "你好": "你好！我是 CWJ AI 助手，有什么可以帮你的吗？",
    "hello": "Hello! I'm CWJ AI Assistant. How can I help you?",
    "你是谁": "我是 CWJ Tools 内置的 AI 聊天机器人。",
    "谢谢": "不客气！",
    "再见": "再见！祝你一天愉快！",
}

def ai_chat(text):
    t = text.lower()
    for k, v in CHAT_RULES.items():
        if k in t:
            return v
    # generic reply
    replies = [
        f"你问的是「{text}」，这是个好问题！我可以帮你把问题整理成搜索关键词。",
        f"关于「{text}」，我建议查阅相关文档或给出更多细节。",
        f"我理解你在问「{text}」，可以试着把问题分成更小的步骤。",
    ]
    return random.choice(replies)

@app.route('/', methods=['GET'])
def index():
    return render_template('aichat.html', chat_input='', chat_result='', chat_error='')

@app.route('/aichat', methods=['POST'])
def aichat():
    ip = request.remote_addr or 'unknown'
    if is_rate_limited(ip):
        logger.warning(f"Rate limit exceeded for {ip}")
        return render_template('aichat.html', chat_input='', chat_result='', chat_error='请求太频繁，请稍后再试')

    raw = request.form.get('chat_input', '')
    text = sanitize_input(raw)
    if not text:
        return render_template('aichat.html', chat_input=escape(raw), chat_result='', chat_error='请输入有效问题（1-500字符）')

    try:
        reply = ai_chat(text)
        return render_template('aichat.html', chat_input=text, chat_result=reply, chat_error='')
    except Exception as e:
        logger.exception('AI chat error')
        return render_template('aichat.html', chat_input=text, chat_result='', chat_error='服务器错误，请稍后重试')

if __name__ == '__main__':
    port = int(os.environ.get('PORT', 5000))
    debug = os.environ.get('DEBUG', 'False').lower() == 'true'
    app.run(host='0.0.0.0', port=port, debug=debug)
